### PR TITLE
[Feature] Support multiple losses during training

### PR DESCRIPTION
**Pull request by MengzhangLI**
Because of being asked frequently by communities, I directly cloned related PR of multiple losses implementation and make a new PR. 

Related PR: https://github.com/open-mmlab/mmsegmentation/pull/244

Related Issues: https://github.com/open-mmlab/mmsegmentation/issues/779, https://github.com/open-mmlab/mmsegmentation/issues/727, https://github.com/open-mmlab/mmsegmentation/issues/486 and so on.

Here is my results on UNet with backbone `UNet-S5-D16` and model `FCN`:

Note: 
(1)  `CE` means its loss function is cross entropy and `DC` means dice loss.
(2) `CE` is cross entropy loss, which is default loss function of MMSegmentation [config](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/unet), I reproduce training to testify real difference between different loss settings.
(3) `loss_weight` is also important. For instance, `(0.5 : 1)` below means the weight of cross entropy loss `CE` and dice loss `DC` is 0.5 and 1, respectively.
(4) I use `--seed 0` but still have some variances from same config setting of different training experiments.

|    Datasets    | CE (from [repo config](https://github.com/open-mmlab/mmsegmentation/tree/master/configs/unet)) | CE (on my own) | CE + DC (1:1) | CE + DC (0.5:1) | CE + DC (1:0.5) | CE + DC (1:3) |
| ---------- | --- | --- | --- | --- | --- | --- |
| DRIVE |  <u> 78.67 </u> | 78.42 | 79.18 | 78.94 | **79.58** | 79.51 |
| STARE       |  81.02 | 81.29 | 82.08 | 81.44 | 82.01 | **82.39** |
| CHASE_DB1       |  80.24 | 80.25 | 80.53 | **80.58** | 80.46 | 80.26 |
| HRF       |  79.45 | 79.24 | 80.66 | 80.72 | 80.6 | **80.79** |

### CODE REVIEW

File: docs/tutorials/training_tricks.md [hunk 1]
```diff
@@ -50,3 +50,19 @@ model=dict(
...
+
+## Multiple Losses
+
+For loss calculation, we support multiple losses training concurrently. Here is an example config of training `unet` on `DRIVE` dataset, whose loss function is `1:3` weighted sum of `CrossEntropyLoss` and `DiceLoss`:
+
+```python
+_base_ = './fcn_unet_s5-d16_64x64_40k_drive.py'
+model = dict(
+    decode_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE', loss_weight=1.0),
+            dict(type='DiceLoss', name='DiCE', loss_weight=3.0)]),
```
Comment (Junjun2016) [inline docs/tutorials/training_tricks.md:62]:
```suggestion
            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
```

File: docs/tutorials/training_tricks.md [hunk 2]
```diff
@@ -50,3 +50,19 @@ model=dict(
...
+    auxiliary_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE',loss_weight=1.0),
+            dict(type='DiceLoss', name='DiCE', loss_weight=3.0)]),
```
Comment (Junjun2016) [inline docs/tutorials/training_tricks.md:64]:
```suggestion
            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
```

File: docs_zh-CN/tutorials/training_tricks.md [hunk 3]
```diff
@@ -49,3 +49,20 @@ model=dict(
...
+## 同时使用多种损失函数 (Multiple Losses)
+
+对于训练时损失函数的计算，我们目前支持多个损失函数同时使用。 以 `unet` 使用 `DRIVE` 数据集训练为例，
+使用 `CrossEntropyLoss` 和 `DiceLoss` 的 `1:3` 的加权和作为损失函数。配置文件写为:
+
+```python
+_base_ = './fcn_unet_s5-d16_64x64_40k_drive.py'
+model = dict(
+    decode_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE', loss_weight=1.0),
+            dict(type='DiceLoss', name='DiCE', loss_weight=3.0)]),
```
Comment (Junjun2016) [inline docs_zh-CN/tutorials/training_tricks.md:62]:
```suggestion
            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
```

File: docs_zh-CN/tutorials/training_tricks.md [hunk 4]
```diff
@@ -49,3 +49,20 @@ model=dict(
...
+    auxiliary_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE',loss_weight=1.0),
+            dict(type='DiceLoss', name='DiCE', loss_weight=3.0)]),
```
Comment (Junjun2016) [inline docs_zh-CN/tutorials/training_tricks.md:64]:
```suggestion
            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
```

File: mmseg/models/decode_heads/decode_head.py [hunk 5]
```diff
@@ -33,8 +33,11 @@ class BaseDecodeHead(BaseModule, metaclass=ABCMeta):
                 a list and passed into decode head.
             None: Only one select feature map is allowed.
             Default: None.
-        loss_decode (dict): Config of decode loss.
-            Default: dict(type='CrossEntropyLoss').
+        loss_decode (dict | Sequence[dict]): Config of decode loss.
+             e.g. dict(type='CrossEntropyLoss'),
+             [dict(type='LossName_1', name='loss_1'),
+              dict(type='LossName_2', name='loss_2')]
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/decode_head.py:39]: Please also add CE and Dice loss to the example.

File: mmseg/models/decode_heads/decode_head.py [hunk 6]
```diff
@@ -33,8 +33,11 @@ class BaseDecodeHead(BaseModule, metaclass=ABCMeta):
...
+            Default: dict(type='CrossEntropyLoss', name='ce').
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/decode_head.py:40]:
```suggestion
            Default: dict(type='CrossEntropyLoss').
```
The default in `__init__`.

File: mmseg/models/decode_heads/decode_head.py [hunk 7]
```diff
@@ -73,9 +76,23 @@ def __init__(self,
...
         self.in_index = in_index
-        self.loss_decode = build_loss(loss_decode)
+
         self.ignore_index = ignore_index
         self.align_corners = align_corners
+        self.loss_decode = nn.ModuleList()
+        self.loss_names = []
+        if isinstance(loss_decode, dict):
+            self.loss_decode.append(build_loss(loss_decode))
+            self.loss_names.append('loss_seg')
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/decode_head.py:86]:
```suggestion
           if 'name' in loss_decode:
                name = loss_decoder.pop('name')
              else:
                 name = 'loss_seg'
            self.loss_decode.append(build_loss(loss_decode))
            self.loss_names.append(name)
```

File: mmseg/models/decode_heads/decode_head.py [hunk 8]
```diff
@@ -73,9 +76,23 @@ def __init__(self,
...
+        elif isinstance(loss_decode, (list, tuple)):
+            for loss in loss_decode:
+                name = loss.pop('name')
+                self.loss_decode.append(build_loss(loss))
+                self.loss_names.append('loss_' + name)
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/decode_head.py:91]:
```suggestion
            for i, loss in enumerate(loss_decode):
            if name in loss:
                name = loss.pop('name')
             else:
                 name = str(i)
                self.loss_decode.append(build_loss(loss))
                self.loss_names.append('loss_' + name)
```

**Comment by caodroid** [comment]
HI， Supporting multiple losses during training is very in need and a nice work，here I also have a suggestion for this work , here
the multiple losses is only for the final prediction, however, if the decoder_head has more than 2 prediction through deep supervisions with differen loss weight, like da_head.py or the network in figure, 
image
 this may not work. for this problem, 



 @force_fp32(apply_to=('seg_logit', ))
    def _losses(self, seg_logit, seg_label, loss_decode):
        """Compute segmentation loss."""
        loss = dict()
        seg_logit = resize(
            input=seg_logit,
            size=seg_label.shape[2:],
            mode='bilinear',
            align_corners=self.align_corners)
        if self.sampler is not None:
            seg_weight = self.sampler.sample(seg_logit, seg_label)
        else:
            seg_weight = None
        seg_label = seg_label.squeeze(1)
      #for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
        loss['loss_seg'] = loss_decode(
            seg_logit,
            seg_label,
            weight=seg_weight,
            ignore_index=self.ignore_index)
        loss['acc_seg'] = accuracy(seg_logit, seg_label)
        return loss

    def losses(self, seg_logit, seg_label):
               """Compute segmentation loss."""
        loss = dict()
        if isinstance(seg_logit, torch.Tensor):   # only one seg_logit 
            for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
                  loss.update(add_prefix( self._losses(seg_logit, seg_label, loss_decode ), loss_name))
       elif isinstance(seg_logit, torch.Tensor):   # multiple seg_logits
            for logit, loss_name, loss_decode in zip(seg_logit, self.loss_names, self.loss_decode):
                   loss.update(add_prefix( self._losses(logit, seg_label, loss_decode ), loss_name))
        return loss

File: mmseg/models/decode_heads/decode_head.py [hunk 9]
```diff
@@ -33,7 +33,10 @@ class BaseDecodeHead(BaseModule, metaclass=ABCMeta):
                 a list and passed into decode head.
             None: Only one select feature map is allowed.
             Default: None.
-        loss_decode (dict): Config of decode loss.
+        loss_decode (dict | Sequence[dict]): Config of decode loss.
+             e.g. dict(type='CrossEntropyLoss'),
+             [dict(type='CrossEntropyLoss', name='CrossEntropy'),
+              dict(type='DiceLoss', name='Dice')]
```
[the remark this answers is not in the document]
Comment (MengzhangLI) [reply mmseg/models/decode_heads/decode_head.py:39]: It has been explained in docs.

**Comment by Junjun2016** [comment]
Should add more unitests to improve the code coverage.

File: docs/tutorials/training_tricks.md [hunk 10]
```diff
@@ -50,3 +50,19 @@ model=dict(
...
+For loss calculation, we support multiple losses training concurrently. Here is an example config of training `unet` on `DRIVE` dataset, whose loss function is `1:3` weighted sum of `CrossEntropyLoss` and `DiceLoss`:
+
+```python
+_base_ = './fcn_unet_s5-d16_64x64_40k_drive.py'
+model = dict(
+    decode_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE', loss_weight=1.0),
+            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
+    auxiliary_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE',loss_weight=1.0),
+            dict(type='DiceLoss', name='Dice', loss_weight=3.0)]),
+    )
```
Comment (Junjun2016) [inline docs/tutorials/training_tricks.md:65]: If the interface of mix loss has been changed, the doc should update accordingly (Chinese and English).

File: mmseg/models/losses/cross_entropy_loss.py [hunk 11]
```diff
@@ -150,14 +150,18 @@ class CrossEntropyLoss(nn.Module):
...
     """
 
     def __init__(self,
                  use_sigmoid=False,
                  use_mask=False,
                  reduction='mean',
                  class_weight=None,
-                 loss_weight=1.0):
+                 loss_weight=1.0,
+                 name='loss_ce'):
```
Comment (Junjun2016) [inline mmseg/models/losses/cross_entropy_loss.py:164]:
```suggestion
                 loss_name='loss_ce'):
```

should be consistent with docstring.

File: mmseg/models/losses/cross_entropy_loss.py [hunk 12]
```diff
@@ -172,6 +176,7 @@ def __init__(self,
             self.cls_criterion = mask_cross_entropy
         else:
             self.cls_criterion = cross_entropy
+        self._loss_name = name
```
Comment (Junjun2016) [inline mmseg/models/losses/cross_entropy_loss.py:179]:
```suggestion
        self._loss_name = loss_name
```

File: mmseg/models/losses/cross_entropy_loss.py [hunk 13]
```diff
@@ -197,3 +202,17 @@ def forward(self,
...
+
+    @property
+    def loss_name(self):
+        """Loss Name.
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
+        name.
```
Comment (Junjun2016) [inline mmseg/models/losses/cross_entropy_loss.py:214]: Should understand the usage of this comment in MMGen.

File: mmseg/models/losses/dice_loss.py [hunk 14]
```diff
@@ -68,6 +68,9 @@ class DiceLoss(nn.Module):
             str format, read them from a file. Defaults to None.
         loss_weight (float, optional): Weight of the loss. Default to 1.0.
         ignore_index (int | None): The label index to be ignored. Default: 255.
+        loss_name (str, optional): Name of the loss item. If you want this loss
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_dice'.
```
Comment (Junjun2016) [inline mmseg/models/losses/dice_loss.py:73]: `loss_` must be the prefix of the name. Defaults to 'loss_dice'.

Why `loss_` must be the prefix of the name in MMGen? and what is the key for MMSeg?

File: mmseg/models/losses/cross_entropy_loss.py [hunk 15]
```diff
@@ -150,14 +150,18 @@ class CrossEntropyLoss(nn.Module):
         class_weight (list[float] | str, optional): Weight of each class. If in
             str format, read them from a file. Defaults to None.
         loss_weight (float, optional): Weight of the loss. Defaults to 1.0.
+        loss_name (str, optional): Name of the loss item. If you want this loss
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_ce'.
```
Comment (Junjun2016) [inline mmseg/models/losses/cross_entropy_loss.py:155]: `loss_` must be the prefix of the name. Defaults to 'loss_dice'.

Why `loss_` must be the prefix of the name in MMGen? and what is the key for MMSeg?

File: mmseg/models/losses/lovasz_loss.py [hunk 16]
```diff
@@ -302,3 +307,17 @@ def forward(self,
...
+
+    @property
+    def loss_name(self):
+        """Loss Name.
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
+        name.
```
Comment (Junjun2016) [inline mmseg/models/losses/lovasz_loss.py:319]: Should understand the usage of this comment in MMGen.

File: mmseg/models/losses/dice_loss.py [hunk 17]
```diff
@@ -118,3 +123,17 @@ def forward(self,
...
+
+    @property
+    def loss_name(self):
+        """Loss Name.
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
+        name.
```
Comment (Junjun2016) [inline mmseg/models/losses/dice_loss.py:135]: Should understand the usage of this comment in MMGen.

File: mmseg/models/losses/dice_loss.py [hunk 18]
```diff
@@ -118,3 +123,17 @@ def forward(self,
...
         return loss
+
+    @property
+    def loss_name(self):
+        """Loss Name.
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
```
Comment (Junjun2016) [inline mmseg/models/losses/dice_loss.py:134]: add more unittests for `mmseg/models/losses/dice_loss.py`.

File: mmseg/models/losses/lovasz_loss.py (hunk 16 above)
Comment (Junjun2016) [reply mmseg/models/losses/lovasz_loss.py:319]: Add more unittest for `mmseg/models/losses/lovasz_loss.py`.

File: mmseg/models/losses/lovasz_loss.py [hunk 19]
```diff
@@ -302,3 +307,17 @@ def forward(self,
             avg_factor=avg_factor,
             **kwargs)
         return loss_cls
+
+    @property
+    def loss_name(self):
```
Comment (Junjun2016) [inline mmseg/models/losses/lovasz_loss.py:312]: Refer to https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/gans/base_gan.py#L41 for understanding.

Comment (Junjun2016) [reply mmseg/models/losses/lovasz_loss.py:312]: Also https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/gans/base_gan.py#L61

Comment (Junjun2016) [reply mmseg/models/losses/lovasz_loss.py:312]: https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/gans/base_gan.py#L71

Comment (Junjun2016) [reply mmseg/models/losses/lovasz_loss.py:312]: https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/gans/base_gan.py#L90

Comment (Junjun2016) [reply mmseg/models/losses/lovasz_loss.py:312]: https://github.com/open-mmlab/mmgeneration/blob/master/mmgen/models/gans/base_gan.py#L251

File: docs_zh-CN/tutorials/training_tricks.md [hunk 20]
```diff
@@ -49,3 +49,20 @@ model=dict(
...
+
+## 同时使用多种损失函数 (Multiple Losses)
+
+对于训练时损失函数的计算，我们目前支持多个损失函数同时使用。 以 `unet` 使用 `DRIVE` 数据集训练为例，
+使用 `CrossEntropyLoss` 和 `DiceLoss` 的 `1:3` 的加权和作为损失函数。配置文件写为:
+
+```python
+_base_ = './fcn_unet_s5-d16_64x64_40k_drive.py'
+model = dict(
+    decode_head=dict(loss_decode=[dict(type='CrossEntropyLoss', name='CE', loss_weight=1.0),
```
Comment (xvjiarui) [inline docs_zh-CN/tutorials/training_tricks.md:61]:
```suggestion
    decode_head=dict(loss_decode=[dict(type='CrossEntropyLoss', loss_name='CE', loss_weight=1.0),
```

File: tests/test_models/test_heads/test_decode_head.py [hunk 21]
```diff
@@ -74,3 +74,62 @@ def test_decode_head():
...
+    # test multi-loss, loss_decode is list of dict
+    inputs = torch.randn(2, 19, 8, 8).float()
+    target = torch.ones(2, 1, 64, 64).long()
+    head = BaseDecodeHead(
+        3,
+        16,
+        num_classes=19,
+        loss_decode=[
+            dict(type='CrossEntropyLoss', name='ce_1'),
+            dict(type='CrossEntropyLoss', name='ce_2')
```
Comment (xvjiarui) [inline tests/test_models/test_heads/test_decode_head.py:106]:
```suggestion
            dict(type='CrossEntropyLoss', loss_name='ce_1'),
            dict(type='CrossEntropyLoss', loss_name='ce_2')
```

File: mmseg/models/decode_heads/decode_head.py [hunk 22]
```diff
@@ -224,10 +238,12 @@ def losses(self, seg_logit, seg_label):
...
-            seg_logit,
-            seg_label,
-            weight=seg_weight,
-            ignore_index=self.ignore_index)
+        for loss_decode in self.loss_decode:
+            loss[loss_decode.loss_name] = loss_decode(
+                seg_logit,
+                seg_label,
+                weight=seg_weight,
+                ignore_index=self.ignore_index)
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/decode_head.py:246]: Add together is `loss_decode.loss_name` already exists.

**Comment by MengzhangLI** [comment]
> HI， Supporting multiple losses during training is very in need and a nice work，here I also have a suggestion for this work , here
> the multiple losses is only for the final prediction, however, if the decoder_head has more than 2 prediction through deep supervisions with differen loss weight, like da_head.py or the network in figure,
> image
> this may not work. for this problem, I think the following code may works
> 
> ```
> @force_fp32(apply_to=('seg_logit', ))
> def _losses(self, seg_logit, seg_label, loss_decode):
>     """Compute segmentation loss."""
>     loss = dict()
>     seg_logit = resize(
>         input=seg_logit,
>         size=seg_label.shape[2:],
>         mode='bilinear',
>         align_corners=self.align_corners)
>     if self.sampler is not None:
>         seg_weight = self.sampler.sample(seg_logit, seg_label)
>     else:
>         seg_weight = None
>     seg_label = seg_label.squeeze(1)
>   #for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
>     loss['loss_seg'] = loss_decode(
>         seg_logit,
>         seg_label,
>         weight=seg_weight,
>         ignore_index=self.ignore_index)
>     loss['acc_seg'] = accuracy(seg_logit, seg_label)
>     return loss
>   
>  def losses(self, seg_logit, seg_label):
>            """Compute segmentation loss."""
>     loss = dict()
>     if isinstance(seg_logit, torch.Tensor):   # only one seg_logit 
>         for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
>               loss.update(add_prefix( self._losses(seg_logit, seg_label, loss_decode ), loss_name))
>    elif isinstance(seg_logit, torch.Tensor):   # multiple seg_logits
>         for logit, loss_name, loss_decode in zip(seg_logit, self.loss_names, self.loss_decode):
>                loss.update(add_prefix( self._losses(logit, seg_label, loss_decode ), loss_name))
>     return loss`
> ```

Thanks for your advice. Deep supervision can be implemented by auxiliary head.

**Comment by caodroid** [comment]
> > HI， Supporting multiple losses during training is very in need and a nice work，here I also have a suggestion for this work , here
> > the multiple losses is only for the final prediction, however, if the decoder_head has more than 2 prediction through deep supervisions with differen loss weight, like da_head.py or the network in figure,
> > image
> > this may not work. for this problem, I think the following code may works
> > ```
> > @force_fp32(apply_to=('seg_logit', ))
> > def _losses(self, seg_logit, seg_label, loss_decode):
> >     """Compute segmentation loss."""
> >     loss = dict()
> >     seg_logit = resize(
> >         input=seg_logit,
> >         size=seg_label.shape[2:],
> >         mode='bilinear',
> >         align_corners=self.align_corners)
> >     if self.sampler is not None:
> >         seg_weight = self.sampler.sample(seg_logit, seg_label)
> >     else:
> >         seg_weight = None
> >     seg_label = seg_label.squeeze(1)
> >   #for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
> >     loss['loss_seg'] = loss_decode(
> >         seg_logit,
> >         seg_label,
> >         weight=seg_weight,
> >         ignore_index=self.ignore_index)
> >     loss['acc_seg'] = accuracy(seg_logit, seg_label)
> >     return loss
> >   
> >  def losses(self, seg_logit, seg_label):
> >            """Compute segmentation loss."""
> >     loss = dict()
> >     if isinstance(seg_logit, torch.Tensor):   # only one seg_logit 
> >         for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
> >               loss.update(add_prefix( self._losses(seg_logit, seg_label, loss_decode ), loss_name))
> >    elif isinstance(seg_logit, torch.Tensor):   # multiple seg_logits
> >         for logit, loss_name, loss_decode in zip(seg_logit, self.loss_names, self.loss_decode):
> >                loss.update(add_prefix( self._losses(logit, seg_label, loss_decode ), loss_name))
> >     return loss`
> > ```
> 
> Thanks for your advice. Deep supervision can be implemented by auxiliary head.

Thanks for your reply,  it's thruth that  the deep supervision can be implemented by auxiliary head, but the  auxiliary head can only supervise the bachbone.  in my advice, I suggest to implemente deep supervision on multiple preditions, for example:  there are three output preditions in da_head, however if I want to change the loss weight on the other auxiliary output predition, it is not achievable in current version.  this demand is very useful for multiple preditions. 
`    def forward(self, inputs):
        """Forward function."""
        x = self._transform_inputs(inputs)
        pam_feat = self.pam_in_conv(x)
        pam_feat = self.pam(pam_feat)
        pam_feat = self.pam_out_conv(pam_feat)
        pam_out = self.pam_cls_seg(pam_feat)

        cam_feat = self.cam_in_conv(x)
        cam_feat = self.cam(cam_feat)
        cam_feat = self.cam_out_conv(cam_feat)
        cam_out = self.cam_cls_seg(cam_feat)

        feat_sum = pam_feat + cam_feat
        pam_cam_out = self.cls_seg(feat_sum)

        return pam_cam_out, pam_out, cam_out`

**Comment by MengzhangLI** [comment]
> > > HI， Supporting multiple losses during training is very in need and a nice work，here I also have a suggestion for this work , here
> > > the multiple losses is only for the final prediction, however, if the decoder_head has more than 2 prediction through deep supervisions with differen loss weight, like da_head.py or the network in figure,
> > > image
> > > this may not work. for this problem, I think the following code may works
> > > ```
> > > @force_fp32(apply_to=('seg_logit', ))
> > > def _losses(self, seg_logit, seg_label, loss_decode):
> > >     """Compute segmentation loss."""
> > >     loss = dict()
> > >     seg_logit = resize(
> > >         input=seg_logit,
> > >         size=seg_label.shape[2:],
> > >         mode='bilinear',
> > >         align_corners=self.align_corners)
> > >     if self.sampler is not None:
> > >         seg_weight = self.sampler.sample(seg_logit, seg_label)
> > >     else:
> > >         seg_weight = None
> > >     seg_label = seg_label.squeeze(1)
> > >   #for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
> > >     loss['loss_seg'] = loss_decode(
> > >         seg_logit,
> > >         seg_label,
> > >         weight=seg_weight,
> > >         ignore_index=self.ignore_index)
> > >     loss['acc_seg'] = accuracy(seg_logit, seg_label)
> > >     return loss
> > >   
> > >  def losses(self, seg_logit, seg_label):
> > >            """Compute segmentation loss."""
> > >     loss = dict()
> > >     if isinstance(seg_logit, torch.Tensor):   # only one seg_logit 
> > >         for loss_name, loss_decode in zip(self.loss_names, self.loss_decode):
> > >               loss.update(add_prefix( self._losses(seg_logit, seg_label, loss_decode ), loss_name))
> > >    elif isinstance(seg_logit, torch.Tensor):   # multiple seg_logits
> > >         for logit, loss_name, loss_decode in zip(seg_logit, self.loss_names, self.loss_decode):
> > >                loss.update(add_prefix( self._losses(logit, seg_label, loss_decode ), loss_name))
> > >     return loss`
> > > ```
> > 
> > 
> > Thanks for your advice. Deep supervision can be implemented by auxiliary head.
> 
> Thanks for your reply, it's thruth that the deep supervision can be implemented by auxiliary head, but the auxiliary head can only supervise the bachbone. in my advice, I suggest to implemente deep supervision on multiple preditions, for example: there are three output preditions in da_head, however if I want to change the loss weight on the other auxiliary output predition, it is not achievable in current version. this demand is very useful for multiple preditions.
> 
> ` def forward(self, inputs):
> 
> ```
>     x = self._transform_inputs(inputs)
>     pam_feat = self.pam_in_conv(x)
>     pam_feat = self.pam(pam_feat)
>     pam_feat = self.pam_out_conv(pam_feat)
>     pam_out = self.pam_cls_seg(pam_feat)
> 
>     cam_feat = self.cam_in_conv(x)
>     cam_feat = self.cam(cam_feat)
>     cam_feat = self.cam_out_conv(cam_feat)
>     cam_out = self.cam_cls_seg(cam_feat)
> 
>     feat_sum = pam_feat + cam_feat
>     pam_cam_out = self.cls_seg(feat_sum)
> 
>     return pam_cam_out, pam_out, cam_out`
> ```

Hi, thanks for your nice proposal and I will think about it in the future, i.e., to make this implementation more flexible just as you mentioned above.

Right now there are already two ways to support it. (1) Like UNet we implemented, the total `encoder-decoder` is the backbone except final fully connected network. (2) Use `NECK` between `BACKBONE` and `HEAD` to handle those feature maps.

Thanks again for your warmhearted proposal.

Best,

File: mmseg/models/losses/dice_loss.py [hunk 23]
```diff
@@ -118,3 +123,17 @@ def forward(self,
...
+        name.
+        Returns:
+            str: The name of this loss item.
+        """
+        return self._loss_name
```
Comment (xvjiarui) [inline mmseg/models/losses/dice_loss.py:139]: Missing unittest

File: mmseg/models/decode_heads/decode_head.py [hunk 24]
```diff
@@ -73,9 +80,20 @@ def __init__(self,
...
         self.align_corners = align_corners
+        self.loss_decode = nn.ModuleList()
+
+        if isinstance(loss_decode, dict):
+            self.loss_decode.append(build_loss(loss_decode))
+        elif isinstance(loss_decode, (list, tuple)):
+            for loss in loss_decode:
+                self.loss_decode.append(build_loss(loss))
+        else:
+            raise TypeError(f'loss_decode must be a dict or sequence of dict,\
```
Comment (xvjiarui) [inline mmseg/models/decode_heads/decode_head.py:94]: Missing unittest

File: mmseg/models/losses/lovasz_loss.py [hunk 25]
```diff
@@ -302,3 +307,17 @@ def forward(self,
...
+
+        This function must be implemented and will return the name of this
+        loss function. This name will be used to combine different loss items
+        by simple sum operation. In addition, if you want this loss item to be
+        included into the backward graph, `loss_` must be the prefix of the
+        name.
+        Returns:
+            str: The name of this loss item.
+        """
+        return self._loss_name
```
Comment (xvjiarui) [inline mmseg/models/losses/lovasz_loss.py:323]: Missing unittest

Comment (xvjiarui) [reply mmseg/models/losses/lovasz_loss.py:323]: We may `assert loss.loss_name == xxx`.

File: mmseg/models/losses/dice_loss.py (hunk 23 above)
Comment (xvjiarui) [reply mmseg/models/losses/dice_loss.py:139]: We may `assert loss.loss_name == xxx`.

**Review by xvjiarui — APPROVED** [review]
LGTM except for the missing unittest

**Comment by Junjun2016** [comment]
Please fix the lint error.

File: mmseg/models/decode_heads/decode_head.py [hunk 26]
```diff
@@ -33,7 +33,14 @@ class BaseDecodeHead(BaseModule, metaclass=ABCMeta):
...
+            The `loss_name` is property of corresponding loss function which
+            could be shown in training log. If you want this loss
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_ce'.
+             e.g. dict(type='CrossEntropyLoss'),
+             [dict(type='CrossEntropyLoss', loss_name='loss_ce'),
+              dict(type='DiceLoss', loss_name='loss_dice')]
             Default: dict(type='CrossEntropyLoss').
         ignore_index (int | None): The label index to be ignored. When using
             masked BCE loss, ignore_index should be set to None. Default: 255
```
Comment (Junjun2016) [inline mmseg/models/decode_heads/decode_head.py:46]:
```suggestion
            masked BCE loss, ignore_index should be set to None. Default: 255.
```

**Review by Junjun2016 — APPROVED** [review]

**Pull request merged**